File: design-add-and-search-words-data-structure/sangbeenmoon.java
import java.util.HashMap;
import java.util.Map;

class WordDictionary {


    Map<Character, WordDictionary> map;
    boolean isEnd = false;

    public WordDictionary() {
        map = new HashMap<>();
    }
    
    public void addWord(String word) {
        char ch = word.charAt(0);
        WordDictionary node;
        if (map.containsKey(ch)) {
            node = map.get(ch);
        } else {
            node = new WordDictionary();
        }

        map.put(ch, node);

        if (word.length() == 1) {
            node.isEnd = true;
            return;    
        }

        node.addWord(word.substring(1));
    }
    
    public boolean search(String word) {
        if (word.length() == 0) {
            return isEnd;
        }

        char ch = word.charAt(0);
        WordDictionary node;

        if (ch == '.') {
            for (Map.Entry<Character, WordDictionary> entry : map.entrySet()) {
                if (entry.getValue().search(word.substring(1))){
                    return true;
                }
            }
        }

        if (map.containsKey(ch)) {
            node = map.get(ch);
            return node.search(word.substring(1));
        }

        return false;
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: design-add-and-search-words-data-structure/sangbeenmoon.py
class WordDictionary:

    def __init__(self):
        self.children = {}
        self.is_end = False

        

    def addWord(self, word: str, i=0) -> None:
        if i >= len(word):
            self.is_end = True
            return

        target_char = word[i]

        if target_char in self.children:
            node = self.children[target_char]
            node.addWord(word, i+1)
        else:
            node = WordDictionary()
            self.children[target_char] = node
            node.addWord(word, i+1)
        

    def search(self, word: str, i=0) -> bool:
        if i >= len(word):
            return self.is_end

        target_char = word[i]

        if target_char == ".":
            for child_node in self.children.values():
                if child_node.search(word,i+1) == True:
                    return True
            
        if target_char in self.children:
            node = self.children[target_char]
            return node.search(word, i+1)
        else:
            node = WordDictionary()
            return node.search(word, i+1)
        
        


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
